File: Sources/waistgunners/Entity.swift
import Foundation

class Entity {
    // base class for basically everything
    var position : Vector
    var size : Vector
    var health : Int

    init(pos : Vector) {
        self.position = pos;
        self.size     = Vector(x : 0, y : 0)
        self.health   = 1
    }

    func update() {
        // whatever it needs to do
        // override in child classes
    }

    func draw() {
        // override in child classes
    }

    func collision(with other : Entity) -> Bool {
        if (self.position.x > other.position.x + other.size.x ||
            self.position.y > other.position.y + other.size.y ||
            other.position.x > self.position.x + self.size.x ||
            other.position.y > self.position.y + self.size.y) {
            return false
        }
        return true
    }
}

File: scripts/Entity.js
/**
 * base class for everything in the game.
 *
 * `update()` and `draw()` should be overriden.
 *
 * overriding `draw_debug()` is optional.
 */
class Entity {
    constructor(position) {
        this.position = position;
        this.size     = new Vector(0, 0);
    }

    update(lapse) {
        // override this
    }

    draw() {
        // override this
    }

    /**
     * detects rectangular collision with another entity
     * @param {Entity} other the other entity with which to detect collision
     * @returns {boolean} returns `true` if the other entity overlaps with this one.
     */
    collision(other) {
        if (other == this) return false;
        if (this.position.x > other.position.x + other.size.x) return false;
        if (this.position.y > other.position.y + other.size.y) return false;
        if (other.position.x > this.position.x + this.size.x) return false;
        if (other.position.y > this.position.y + this.size.y) return false;
        return true;
    }

    draw_debug() {

    }
}